### PR TITLE
[WIP] NormaliseEvent pipe

### PR DESCRIPTION
**Pull request by rhiaro**
This sets up the NormaliseEvent pipe which can handle flat Events/OnDemandEvents and anything with subEvents of type Event/OnDemandEvent.

Also has a lot of the boilerplate needed for the other pipes.

WIP:

* provenanceInformation is not filled in yet
* empty values and invalid properties are not corrected yet 
* ~~there's a bug in the normalise-data pipe runner where it gets stuck in the while loop if there are no normalisedEvents to store (I think)~~

(current failing tests are expected due to first two points)

### CODE REVIEW

**Comment by rhiaro** [comment]
[Normalisations](https://trello.com/c/C70upiyi/10-normalisations)

File: src/lib/normalise-data.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import pipes from './pipes/index.js';
 async function normalise_data_all_publisher_feeds() {
   const client = await database_pool.connect();
     try {
-        const res_find_publisher_feeds = await client.query('SELECT * FROM publisher_feed');
+        const res_find_publisher_feeds = await client.query('SELECT * FROM publisher_feed WHERE');
```
Comment (odscjames) [inline src/lib/normalise-data.js:9]: WHERE is I think left over debugging?

File: src/lib/pipes/index.js [hunk 2]
```diff
@@ -1,6 +1,8 @@
-//import TestPipe from './test-pipe.js';
+import TestPipe from './test-pipe.js';
```
Comment (odscjames) [inline src/lib/pipes/index.js:1]: Also a debug thing, but it's just an import so maybe it's fine

File: src/lib/normalise-data.js (hunk 1 above)
Comment (rhiaro) [reply src/lib/normalise-data.js:9]: oops well spotted :p

File: src/lib/pipes/normalise-event-pipe.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        console.log(`Running ${id} (${type}) through ${this.constructor.name}`);
+
+        if (type == 'Event' || type == 'OnDemandEvent'){
+            // The top level event is the Event
+
+            this.doCleanup();
+
+            if (typeof this.rawData.superEvent !== 'undefined'){
+                // It has a superEvent which we can get more data from.
+                let {superEvent, ...event} = this.rawData;
```
Comment (odscjames) [inline src/lib/pipes/normalise-event-pipe.js:23]: I don't understand what the ... syntax does on the left hand side of an assign statement?

File: src/lib/pipes/pipe.js [hunk 4]
```diff
@@ -1,20 +1,251 @@
...
+    // rawMeta is an entire database row from the raw_data table, minus data field
+    // rawData is the contents of the data field
+    const { data, ...meta } = rawData;
+    this.rawMeta = meta;
+    if(data == null){
+      this.rawData = {};
+    }else{
+      this.rawData = data;
+    }
+    this.provenance = {};
```
Comment (odscjames) [inline src/lib/pipes/pipe.js:16]: Said we'd take this out for now

File: src/lib/pipes/normalise-event-pipe.js (hunk 3 above)
Comment (rhiaro) [reply src/lib/pipes/normalise-event-pipe.js:23]: It assigns the superEvent property on the rawData object to a variable called superEvent, and all the rest of the object to the variable called event.

Comment (odscjames) [reply src/lib/pipes/normalise-event-pipe.js:23]: ah ok

File: src/lib/utils.js [hunk 5]
```diff
@@ -22,8 +25,31 @@ class Utils {
...
+      return JSON.parse(
+        await fs.readFile(path, {encoding: 'utf8'})
+      );
+    } catch (error){
+        console.error(error);
+    }
+  }
+
+  static async getContext(){
+    return await Utils.readJson(path.resolve(path.resolve(), './src/lib/oa.jsonld'));
```
Comment (odscjames) [inline src/lib/utils.js:39]: Isn't there a website we can load this from? But we can put this in for now.

Comment (rhiaro) [reply src/lib/utils.js:39]: it seemed more efficient to cache it locally. It's also better practice as every network request is an attack vector :)

File: test/test-pipe.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                "https://schema.org/",
+            ];
+
+            let pipe1 = new NormaliseEventPipe({id: 1, data: {"@context": badContexts1}}, []);
+            pipe1.fixContext();
+            assert.deepEqual(pipe1.rawData["@context"], goodContexts1);
+
+            let pipe2 = new NormaliseEventPipe({id: 2, data: {"@context": badContexts2}}, []);
+            pipe2.fixContext();
+            assert.deepEqual(pipe2.rawData["@context"], goodContexts2);
```
Comment (odscjames) [inline test/test-pipe.js:50]: I would have put different asserts like this into different tests (eg it() blocks). With only one test, if the first assert fails you don't see the results of the latter asserts and that might hide a failure. If they were different tests you'd see each failure each time, which is nice. But not important enough to fix now.

Comment (rhiaro) [reply test/test-pipe.js:50]: okie dokie, noted for next batch of updates

File: src/lib/pipeline.js [hunk 7]
```diff
@@ -17,9 +17,8 @@ class PipeLine {
           console.log(`Error running data through pipe ${pipeSection.constructor.name} \n ${error}`);
         }
       }
-      if (normalisedEvents) {
-        await this.pipeOutputCb( this.rawData.id, normalisedEvents,);
-      }
+
```
Comment (odscjames) [inline src/lib/pipeline.js:20]: I would have put a comment here

// We always want to call pipeOutputCb even if normalisedEvents is empty so pipeOutputCb will record that we have checked this bit of data

Just stops someone trying to add it back later without realising there is a good reason

But maybe later

File: src/lib/pipes/pipe.js (hunk 4 above)
Comment (odscjames) [inline src/lib/pipes/pipe.js]: What is this doing if we add to raw_data anyway? (Not totally sure adding to raw_data is best way anyway, maybe have some extra fields you pass to NormalisedEvent class? Can leave in for now, but I would have taken this and provanence function out totally

Comment (rhiaro) [reply src/lib/pipes/pipe.js]: Yeah sorry this was leftover from a previous slightly different way of doing the provenance. Taken it out with recent commit.

File: src/lib/utils.js (hunk 5 above)
Comment (odscjames) [reply src/lib/utils.js:39]: We can load once at startup then cache that (like we did for org tree last time), but yeah, it's fine for now

File: src/lib/utils.js [hunk 8]
```diff
@@ -22,8 +25,31 @@ class Utils {
...
+  }
 
+  static ensureArray(input){
```
Comment (odscjames) [inline src/lib/utils.js:42]: Won't ensureArray(undefined) return [ensureArray] where you probably want it to return []?

Comment (rhiaro) [reply src/lib/utils.js:42]: I will write a test for and then fix that in a new PR :)

**Pull request merged**